pkg/storegateway/bucket_stores.go

```diff
@@ -151,11 +152,30 @@ func (u *BucketStores) InitialSync(ctx context.Context) error {
 
 // SyncBlocks synchronizes the stores state with the Bucket store for every user.
 func (u *BucketStores) SyncBlocks(ctx context.Context) error {
-	return u.syncUsersBlocks(ctx, func(ctx context.Context, s *store.BucketStore) error {
+	return u.syncUsersBlocksWithRetries(ctx, func(ctx context.Context, s *store.BucketStore) error {
 		return s.SyncBlocks(ctx)
 	})
 }
 
+func (u *BucketStores) syncUsersBlocksWithRetries(ctx context.Context, f func(context.Context, *store.BucketStore) error) error {
+	retries := util.NewBackoff(ctx, util.BackoffConfig{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: 10 * time.Second,
+		MaxRetries: 3,
+	})
+
+	for retries.Ongoing() {
+		err := u.syncUsersBlocks(ctx, f)
+		if err == nil {
+			return err
+		}
+
+		retries.Wait()
+	}
+
+	return retries.Err()
```

Comment:
This would suppress the actual error. I would save the last error returned by `u.syncUsersBlocks(ctx, f)` and return it once all retries expired (eg. see `joinMembersOnStartup()` at `pkg/ring/kv/memberlist/memberlist_client.go` for reference).